MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java

```diff
@@ -0,0 +1,316 @@
+
+import com.adobe.internal.xmp.XMPException;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+/**
+ * An Android app widget to set up a video player for a motion photo file.
+ */
+public class MotionPhotoWidget extends TextureView {
+
+    private static final String TAG = "MotionPhotoWidget";
+
+    private final boolean autoloop;
+    private ExecutorService executor;
+    private MotionPhotoReader reader;
+    private boolean isPaused = true;
+    private String filename;
+    private SurfaceTexture savedSurfaceTexture;
+    private PlayProcess playProcess;
+
+    /** Fields that are saved for the view state. */
+    private long savedTimestampUs;
+    private int videoWidth;
+    private int videoHeight;
+    private int viewWidth;
+    private int viewHeight;
+
+    public MotionPhotoWidget(Context context) {
+        super(context);
+        autoloop = true;
+        setup();
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    public MotionPhotoWidget(Context context, @Nullable AttributeSet attrs) {
+        super(context, attrs);
+        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.MotionPhotoWidget,
+                /* defStyleAttr = */ 0,
+                /* defStyleRes = */ 0);
+
+        // Fetch value of “custom:background_color”
+        Color backgroundColor = Color.valueOf(
+                ta.getColor(R.styleable.MotionPhotoWidget_background_color, Color.RED)
+        );
+
+        // Fetch value of “custom:timeline_base_color”
+        Color timelineBaseColor = Color.valueOf(
+                ta.getColor(R.styleable.MotionPhotoWidget_timeline_base_color, Color.RED)
+        );
+
+        // Fetch value of “custom:timeline_fill_color”
+        Color timelineFillColor = Color.valueOf(
+                ta.getColor(R.styleable.MotionPhotoWidget_timeline_fill_color, Color.RED)
+        );
+
+        // Fetch value of “custom:autoloop”
+        autoloop = ta.getBoolean(R.styleable.MotionPhotoWidget_autoloop, true);
+        ta.recycle();
+        setup();
+    }
+
+    /**
+     * Sets up the executor, the play/pause process to be executed, and the surface texture
+     * listener. This should only be called in a constructor, and should be called in every
+     * constructor.
+     */
+    private void setup() {
+        // Set up the executor and play/pause process to facilitate stopping and starting the video
+        playProcess = new PlayProcess();
+        executor = Executors.newSingleThreadExecutor(new ThreadFactory() {
+            @Override
+            public Thread newThread(Runnable r) {
+                return new Thread(r) {
+                    @Override
+                    public void interrupt() {
+                        super.interrupt();
+                        playProcess.cancel();
+                    }
+                };
+            }
+        });
```

Comment:
This is an interesting idea but might be overkill for this. You should just directly call `playProcess.cancel()` in the associated close/shutdown (which is also missing).

This executor will leak everytime it is created. Also, `executor.interrupt` nor `executor.shutdown`  is never called, so this thread will never be interrupted. Consider either overriding the view's `onDetachFromWindow` and `onAttachToWindow` to setup the begin and end lifecylce.

A cleaner approach might be to migrate to `Fragment` since those do properly implement a create/destroy lifecycle